extra_data/keydata.py

```diff
@@ -0,0 +1,153 @@
+import numpy as np
+
+class KeyData:
+    def __init__(self, source, key, data_chunks, section):
+        self.source = source
+        self.key = key
+        self.section = section
+
+        dim0 = 0
+        entry_shapes, dtypes = set(), set()
+        for chunk in data_chunks:
+            dim0 += chunk.total_count
+            entry_shapes.add(chunk.dataset.shape[1:])
+            dtypes.add(chunk.dataset.dtype)
+
+        if len(entry_shapes) > 1:
+            raise Exception("Mismatched data shapes: {}".format(entry_shapes))
+        if len(dtypes) > 1:
+            raise Exception("Mismatched dtypes: {}".format(dtypes))
+
+        self.dtype = dtypes.pop()
+        self.entry_shape = entry_shapes.pop()
+        self.shape = (dim0,) + self.entry_shape
+        self.ndim = len(self.shape)
+        # Sort chunks by train ID and discard any empty ones
+        self._data_chunks = sorted(
+            [c for c in data_chunks if c.total_count > 0],
+            key=lambda c: c.train_ids[0]
+        )
+        if self._data_chunks:
+            self.train_ids = np.concatenate([c.train_ids for c in self._data_chunks])
+        else:
+            self.train_ids = np.zeros(shape=(0,), dtype=np.uint64)
+
+    def __repr__(self):
+        return f"<extra_data.KeyData source={self.source!r} key={self.key!r} " \
+               f"for {len(self.train_ids)} trains"
+
+    @property
+    def _key_group(self):
+        """The part of the key needed to look up index data"""
+        if self.section == 'INSTRUMENT':
+            return self.key.partition('.')[0]
+        else:
+            return ''
+
+    @property
+    def _data_path(self):
```

Comment:
Why advertise this property as private?
I couldn't find any internal use for it, however, at the same time it might be useful to have at times.

Reply:
I've got into the habit of marking things private unless I've got a good justification for making them public - a smaller public API means fewer distractions for users working out what they need, and makes changes easier in the future, because it's more reasonable to break private APIs. But it's not necessarily a good thing overall, and we could make this public.

The lack of internal use is related to my third question above - should `run[source, key]` find all data chunks, or hand `KeyData` the files and let it look up chunks on demand? I started with the latter, switched to the former, and may yet switch back.

Reply:
This is now public as `hdf5_data_path`.